tests/test_grid3d/test_grid_grdecl.py

```diff
@@ -0,0 +1,366 @@
+    nx, ny, nz = grdecl_grid.dimensions
+    assert_allclose(grdecl_grid.xtgeo_zcorn(), xtggrid._zcornsv, atol=0.02)
+
+
+@given(xtgeo_compatible_grdecl_grids)
+def test_to_from_grdeclgrid(grdecl_grid):
+    xtggrid = Grid()
+    xtggrid._actnumsv = grdecl_grid.xtgeo_actnum()
+    xtggrid._coordsv = grdecl_grid.xtgeo_coord()
+    xtggrid._zcornsv = grdecl_grid.xtgeo_zcorn()
+    nx, ny, nz = grdecl_grid.dimensions
+    xtggrid._ncol = nx
+    xtggrid._nrow = ny
+    xtggrid._nlay = nz
+    xtggrid._xtgformat = 2
+
+    grdecl_grid2 = ggrid.GrdeclGrid.from_xtgeo_grid(xtggrid)
+    assert_allclose(grdecl_grid2.xtgeo_actnum(), xtggrid._actnumsv, atol=0.02)
+    assert_allclose(grdecl_grid2.xtgeo_coord(), xtggrid._coordsv, atol=0.02)
+    assert_allclose(grdecl_grid2.xtgeo_zcorn(), xtggrid._zcornsv, atol=0.02)
+
+
+@settings(deadline=None, suppress_health_check=[HealthCheck.function_scoped_fixture])
+@given(xtgeo_compatible_grdecl_grids)
+def test_to_from_xtggrid_write(tmp_path, grdecl_grid):
+    xtggrid = Grid()
+    xtggrid._actnumsv = grdecl_grid.xtgeo_actnum()
+    xtggrid._coordsv = grdecl_grid.xtgeo_coord()
+    xtggrid._zcornsv = grdecl_grid.xtgeo_zcorn()
+    nx, ny, nz = grdecl_grid.dimensions
+    xtggrid._ncol = nx
+    xtggrid._nrow = ny
+    xtggrid._nlay = nz
+    xtggrid._xtgformat = 2
+
+    xtggrid.to_file(tmp_path / "xtggrid.grdecl", fformat="grdecl")
+    grdecl_grid2 = ggrid.GrdeclGrid.from_file(tmp_path / "xtggrid.grdecl")
+
+    assert_allclose(grdecl_grid.zcorn, grdecl_grid2.zcorn, atol=0.1)
+    assert_allclose(grdecl_grid.xtgeo_coord(), grdecl_grid2.xtgeo_coord(), atol=0.1)
+    if grdecl_grid.actnum is None:
+        assert grdecl_grid2.actnum is None or np.all(grdecl_grid2.actnum)
+    else:
+        assert np.array_equal(grdecl_grid.actnum, grdecl_grid2.actnum)
+
+
+@settings(
+    deadline=None,
+    suppress_health_check=[HealthCheck.function_scoped_fixture],
+)
+@given(xtgeo_compatible_grdecl_grids)
+def test_from_to_grdeclgrid_write(tmp_path, grdecl_grid):
+    xtggrid = Grid()
+
+    grdecl_grid.to_file(tmp_path / "xtggrid.grdecl")
+    xtggrid = Grid(tmp_path / "xtggrid.grdecl", fformat="grdecl")
+
+    xtggrid._xtgformat2()
+    assert_allclose(grdecl_grid.xtgeo_zcorn(), xtggrid._zcornsv, atol=0.1)
+    assert_allclose(grdecl_grid.xtgeo_coord(), xtggrid._coordsv, atol=0.02)
+    assert_allclose(grdecl_grid.xtgeo_zcorn(), xtggrid._zcornsv, atol=0.02)
+
+
+@given(xtgeo_compatible_grdecl_grids)
+def test_xtgeo_values_are_c_contiguous(grdecl_grid):
+    assert grdecl_grid.xtgeo_coord().flags["C_CONTIGUOUS"]
+    assert grdecl_grid.xtgeo_actnum().flags["C_CONTIGUOUS"]
+    assert grdecl_grid.xtgeo_zcorn().flags["C_CONTIGUOUS"]
+
+
+@given(grdecl_grids())
+def test_eq_reflexivity(grdecl_grid):
+    assert grdecl_grid == grdecl_grid
+
+
+@given(grdecl_grids(), grdecl_grids())
+def test_eq_symmetry(grdecl_grid1, grdecl_grid2):
+    if grdecl_grid1 == grdecl_grid2:
+        assert grdecl_grid2 == grdecl_grid1
```

Comment:
An incorrectly implemented eq function could have this invalidated yes. It would require some funky corner cases, but say you have optional fields for instance:
```
@dataclass
class Person:
   name: str
   age: Optional[int] = None
   
   def __eq__(this, other):
        is_eq = this.name == other.name
        if this.age is not None:
            return is_eq and this.age == other.age
    ```
    this does not pass the symmetry test as `Person("Paul", 32) != Person("Paul", None)` but `Person("Paul", None) == Person("Paul", 32)`.